debugger.md

```diff
@@ -0,0 +1,160 @@
+# Debugger proposal
+
+## Background
+
+One of the greatest advantages of being a log based workflow engine is that Temporal workflows are replayable.
+
+A developer can download workflow history and run it through the SDK replayer to materialize the workflow program's
+state.
+
+Workflows, being "just code", can be debugged with tools the developer is familiar with.
+
+## Features
+
+### Overview
+
+Typically, when looking at workflow history, it's clear which workflow task a failure occurs in but it's not immediately
+obvious which line of code that correlates to.
+
+This proposal discusses how we can make the debugging experience better for developers.
+
+### Supported languages
+
+Initially we will support Core based languages (TypeScript and Python) but Go and Java could also be supported with
+additional effort.
+
+The main advantage of going for Core based languages is that those language SDKs run Rust Core in a separate thread
+pool, allowing interaction with the runtime while a workflow is paused by a debugger.
+
+Targeting Core based SDKs will allow us to reuse a lot of the code for Ruby, .NET and possibly other SDKs if we decide
+to rewrite them to be backed by Core.
+
+### Supported IDEs
+
+Initially we will support Microsoft's Visual Studio Code only while keeping in mind other IDEs.
+
+The rationale being that at least for JavaScript VSCode is by far the most popular IDE and we'd get the most value
+investing in providing good DX for that environment.
+
+### Components
+
+- VSCode plugin
+  - Main panel / Workflow history panel
+  - Debug adapter
+- Core replay debug server
+- Lang middleware
+
+#### Main panel
+
+Has input to enter workflow ID and optional run ID with submit button to start a debug session.
+Alternatively can have a file (path) input to use a history JSON or binary file (as exported by web and tctl or
+downloaded with a client).
+
+The credentials for connecting to a server should be stored in the system keychain (can use
```

Comment:
Just researched, there's a vscode secrets API for this: https://code.visualstudio.com/api/references/vscode-api#ExtensionContext.secrets (commit: https://github.com/microsoft/vscode/commit/d3965a2b4dd54f1a415fa370fea90b84bc7d9b7c, associated PR specifically mentions so extension authors don't need keytar)

Also may need a bit info on where/how the server address and creds are inputted by the user. Also would suggest supporting in-vscode-config creds if they so choose. No detail needed, just want to know how I point to my server. Is there a vscode "servers" concept that can be leveraged?

Reply:
Thanks for the tip, will change to use builtin secrets.

Reply:
I was thinking we'd have a form in the extension panel to configure the server settings.
I'll add to the proposal.